packages/api-client-core/src/exchanges/cacheExchange.ts

```diff
@@ -0,0 +1,237 @@
+          cachedResult ||
+          makeResult(operation, {
+            data: null,
+          });
+
+        result = {
+          ...result,
+          operation: addMetadata(operation, {
+            cacheOutcome: cachedResult ? "hit" : "miss",
+          }),
+        };
+
+        if (operation.context.requestPolicy === "cache-and-network") {
+          result.stale = true;
+          reexecuteOperation(client, operation);
+        }
+
+        return result;
+      })
+    );
+
+    const forwardedOps$ = pipe(
+      merge([
+        pipe(
+          ops$,
+          filter((op) => !shouldSkip(op) && !isOperationCached(op)),
+          map(mapTypeNames)
+        ),
+        pipe(
+          ops$,
+          filter((op) => shouldSkip(op))
+        ),
+      ]),
+      map((op) => addMetadata(op, { cacheOutcome: "miss" })),
+      filter((op) => op.kind !== "query" || op.context.requestPolicy !== "cache-only"),
+      forward,
+      tap((response) => {
+        let { operation } = response;
+        if (!operation) return;
+
+        let typenames = operation.context.additionalTypenames || [];
+        // NOTE: For now, we only respect `additionalTypenames` from subscriptions to
+        // avoid unexpected breaking changes
+        // We'd expect live queries or other update mechanisms to be more suitable rather
+        // than using subscriptions as “signals” to reexecute queries. However, if they’re
+        // just used as signals, it’s intuitive to hook them up using `additionalTypenames`
+        if (response.operation.kind !== "subscription") {
+          typenames = collectTypenames(response.data).concat(typenames);
+        }
+
+        // Invalidates the cache given a mutation's response
+        if (response.operation.kind === "mutation" || response.operation.kind === "subscription") {
+          const pendingOperations = new Set<number>();
+
+          dispatchDebug({
+            type: "cacheInvalidation",
+            message: `The following typenames have been invalidated: ${typenames}`,
+            operation,
+            data: { typenames, response },
+          });
+
+          for (let i = 0; i < typenames.length; i++) {
+            const typeName = typenames[i];
+            let operations = operationCache.get(typeName);
+            if (!operations) operationCache.set(typeName, (operations = new Set()));
+            for (const key of operations.values()) pendingOperations.add(key);
+            operations.clear();
+          }
+
+          for (const key of pendingOperations.values()) {
+            if (resultCache.has(key)) {
+              operation = (resultCache.get(key) as OperationResult).operation;
+              // GADGET: added the below line to skip reexecuting live queries when their data is invalidated since they should receive updates from the server
+              if (!operation.query.definitions.some(isLiveQueryOperationDefinitionNode)) {
+                resultCache.delete(key);
```

Comment:
Should we still delete, or is there potentially a race condition? Just wondering if there's any opportunity for stale cache.